File: Go/examples/csv_utils_example.go
// Example program demonstrating CSV utilities
package main

import (
	"fmt"
	"log"
	"strings"

	"github.com/ayukyo/alltoolkit/Go/csv_utils"
)

func main() {
	fmt.Println("=== CSV Utils Example ===\n")

	// Example 1: Reading CSV from string
	fmt.Println("1. Reading CSV from string:")
	csvData := `name,age,city,salary
Alice,30,New York,75000
Bob,25,Los Angeles,60000
Charlie,35,Chicago,90000
Diana,28,Seattle,72000`

	data, err := csv_utils.ReadString(csvData)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Headers: %v\n", data.Headers)
	fmt.Printf("Row count: %d\n\n", data.RowCount())

	// Example 2: Accessing data with type conversion
	fmt.Println("2. Accessing data with type conversion:")
	for _, row := range data.Rows {
		name := row.Get("name")
		age := row.GetInt("age", 0)
		city := row.Get("city")
		salary := row.GetFloat("salary", 0.0)
		fmt.Printf("  %s is %d years old, lives in %s, earns $%.2f\n", name, age, city, salary)
	}
	fmt.Println()

	// Example 3: Filtering rows
	fmt.Println("3. Filtering rows (age > 28):")
	olderThan28 := data.FilterRows(func(row csv_utils.CsvRow) bool {
		return row.GetInt("age", 0) > 28
	})
	for _, row := range olderThan28.Rows {
		fmt.Printf("  %s: %d years old\n", row.Get("name"), row.GetInt("age", 0))
	}
	fmt.Println()

	// Example 4: Sorting by column
	fmt.Println("4. Sorting by salary (ascending):")
	sortedBySalary := data.SortByWithOptions("salary", true, true)
	for _, row := range sortedBySalary.Rows {
		fmt.Printf("  %s: $%.2f\n", row.Get("name"), row.GetFloat("salary", 0.0))
	}
	fmt.Println()

	// Example 5: Column statistics
	fmt.Println("5. Column statistics:")
	fmt.Printf("  Average salary: $%.2f\n", data.AvgColumn("salary"))
	fmt.Printf("  Min salary: $%.2f\n", data.MinColumn("salary"))
	fmt.Printf("  Max salary: $%.2f\n", data.MaxColumn("salary"))
	fmt.Printf("  Total salary: $%.2f\n\n", data.SumColumn("salary"))

	// Example 6: Filtering columns
	fmt.Println("6. Filtering columns (name and city only):")
	nameCityOnly := data.FilterColumns([]string{"name", "city"})
	for _, row := range nameCityOnly.Rows {
		fmt.Printf("  %s lives in %s\n", row.Get("name"), row.Get("city"))
	}
	fmt.Println()

	// Example 7: Grouping by column
	fmt.Println("7. Grouping by city:")
	byCity := data.GroupBy("city")
	for city, rows := range byCity {
		names := []string{}
		for _, row := range rows {
			names = append(names, row.Get("name"))
		}
		fmt.Printf("  %s: %v\n", city, names)
	}
	fmt.Println()

	// Example 8: Finding rows
	fmt.Println("8. Finding rows:")
	bob := data.Find(func(row csv_utils.CsvRow) bool {
		return row.Get("name") == "Bob"
	})
	if bob != nil {
		fmt.Printf("  Found Bob: %s\n", bob.Get("city"))
	}

	highEarners := data.FindAll(func(row csv_utils.CsvRow) bool {
		return row.GetFloat("salary", 0) > 70000
	})
	fmt.Printf("  High earners (> $70k): %d people\n", len(highEarners))
	fmt.Println()

	// Example 9: Writing CSV
	fmt.Println("9. Writing CSV:")
	writer := csv_utils.NewWriter()
	writer.SetHeaders([]string{"product", "price", "quantity"})
	writer.AddRow([]string{"Laptop", "999.99", "5"})
	writer.AddRow([]string{"Mouse", "29.99", "20"})
	writer.AddRow([]string{"Keyboard", "79.99", "10"})

	csvString := writer.ToString()
	fmt.Println("  Generated CSV:")
	fmt.Println(csvString)

	// Example 10: Custom delimiter (TSV)
	fmt.Println("10. Reading TSV (tab-separated values):")
	tsvData := `name\tage\tcity
Alice\t30\tNew York
Bob\t25\tLos Angeles`

	opts := csv_utils.DefaultOptions()
	opts.Delimiter = '\t'
	tsv, err := csv_utils.ReadStringWithOptions(tsvData, opts)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("  TSV Headers: %v\n", tsv.Headers)
	fmt.Printf("  First row: %s, %s, %s\n\n",
		tsv.Rows[0].Get("name"),
		tsv.Rows[0].Get("age"),
		tsv.Rows[0].Get("city"))

	// Example 11: Detecting delimiter
	fmt.Println("11. Detecting delimiter:")
	sample1 := "a,b,c"
	sample2 := "a;b;c"
	fmt.Printf("  Delimiter in '%s': %c\n", sample1, csv_utils.DetectDelimiter(sample1))
	fmt.Printf("  Delimiter in '%s': %c\n\n", sample2, csv_utils.DetectDelimiter(sample2))

	// Example 12: Distinct values
	fmt.Println("12. Distinct values:")
	csvWithDuplicates := `category,value
A,10
B,20
A,30
C,40
B,50`

	data2, _ := csv_utils.ReadString(csvWithDuplicates)
	distinctCategories := data2.Distinct("category")
	fmt.Printf("  Distinct categories: %v\n\n", distinctCategories)

	// Example 13: Transforming data
	fmt.Println("13. Transforming data (uppercase names):")
	data.Transform(func(row csv_utils.CsvRow) csv_utils.CsvRow {
		row["name"] = strings.ToUpper(row["name"])
		return row
	})
	for _, row := range data.Rows {
		fmt.Printf("  %s\n", row.Get("name"))
	}
	fmt.Println()

	// Example 14: Joining datasets
	fmt.Println("14. Joining datasets:")
	left := &csv_utils.CsvData{
		Headers: []string{"id", "name"},
		Rows: []csv_utils.CsvRow{
			{"id": "1", "name": "Alice"},
			{"id": "2", "name": "Bob"},
		},
	}
	right := &csv_utils.CsvData{
		Headers: []string{"department", "salary"},
		Rows: []csv_utils.CsvRow{
			{"department": "Engineering", "salary": "90000"},
			{"department": "Sales", "salary": "70000"},
		},
	}
	joined, err := csv_utils.Join(left, right)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("  Joined headers: %v\n", joined.Headers)
	for _, row := range joined.Rows {
		fmt.Printf("  %s works in %s, earns $%s\n",
			row.Get("name"), row.Get("department"), row.Get("salary"))
	}
	fmt.Println()

	// Example 15: Validating CSV
	fmt.Println("15. Validating CSV:")
	validCsv := "a,b\n1,2"
	invalidCsv := "a,b\n1,2,3" // Mismatched columns (will parse but may not be what expected)
	fmt.Printf("  Is '%s' valid? %v\n", validCsv, csv_utils.IsValidCsv(validCsv))
	fmt.Printf("  Is '%s' valid? %v\n\n", invalidCsv, csv_utils.IsValidCsv(invalidCsv))

	// Example 16: Boolean conversion
	fmt.Println("16. Boolean conversion:")
	csvWithBools := `name,active,admin
Alice,true,yes
Bob,false,no
Charlie,1,0`

	data3, _ := csv_utils.ReadString(csvWithBools)
	for _, row := range data3.Rows {
		name := row.Get("name")
		active := row.GetBool("active", false)
		admin := row.GetBool("admin", false)
		fmt.Printf("  %s: active=%v, admin=%v\n", name, active, admin)
	}
	fmt.Println()

	// Example 17: Counting with predicate
	fmt.Println("17. Counting with predicate:")
	count := data.Count(func(row csv_utils.CsvRow) bool {
		return row.GetFloat("salary", 0) > 70000
	})
	fmt.Printf("  Employees earning more than $70k: %d\n\n", count)

	// Example 18: Adding and removing columns
	fmt.Println("18. Adding and removing columns:")
	dataCopy := &csv_utils.CsvData{
		Headers: []string{"name", "age"},
		Rows: []csv_utils.CsvRow{
			{"name": "Alice", "age": "30"},
			{"name": "Bob", "age": "25"},
		},
	}
	dataCopy.AddColumn("country", []string{"USA", "UK"})
	fmt.Printf("  After adding 'country' column: %v\n", dataCopy.Headers)
	dataCopy.RemoveColumn("age")
	fmt.Printf("  After removing 'age' column: %v\n", dataCopy.Headers)
	fmt.Println()

	fmt.Println("=== End of Examples ===")
}

File: TypeScript/examples/csv_utils_example.ts
/**
 * CSV Utilities Example
 * 
 * This file demonstrates the usage of the CSV Utilities module.
 * 
 * To run this example:
 *   cd TypeScript/examples
 *   npx ts-node csv_utils_example.ts
 */

import {
  parse,
  stringify,
  isValidCsv,
  getStats,
  filterRows,
  selectColumns,
  sortBy,
  addColumn,
  removeColumn,
  renameColumn,
  mergeVertical,
  detectDelimiter,
  toJson,
  fromJson,
} from '../csv_utils/mod';

console.log('=== CSV Utilities Examples ===\n');

// Example 1: Parse CSV string
console.log('1. Parse CSV string:');
const csvString = `
name,age,city
John Doe,30,New York
Jane Smith,25,Los Angeles
Bob Johnson,35,Chicago
Alice Brown,28,San Francisco
`;

const parsed = parse(csvString);
console.log('Headers:', parsed.headers);
console.log('First row:', parsed.rows[0]);
console.log('');

// Example 2: Parse CSV with quotes and special characters
console.log('2. Parse CSV with quotes:');
const csvWithQuotes = `
name,description,price
"Doe, John","A software developer, who loves coding",50000
Jane Smith,"Senior Developer",75000
`;

const parsedQuotes = parse(csvWithQuotes);
console.log('Parsed with quotes:', parsedQuotes.rows);
console.log('');

// Example 3: Convert data to CSV
console.log('3. Convert data to CSV:');
const data = [
  { product: 'Laptop', price: 999.99, stock: 50 },
  { product: 'Mouse', price: 29.99, stock: 200 },
  { product: 'Keyboard', price: 79.99, stock: 100 },
];

const csvOutput = stringify(data);
console.log('Generated CSV:');
console.log(csvOutput);

// Example 4: CSV with custom delimiter
console.log('4. CSV with semicolon delimiter:');
const semicolonData = stringify(data, { delimiter: ';' });
console.log(semicolonData);

// Example 5: Validate CSV
console.log('5. Validate CSV:');
const validCsv = 'name,age\nJohn,30';
const invalidCsv = 'name,age\nJohn,30\n"unclosed quote';
console.log('Is valid CSV:', isValidCsv(validCsv));
console.log('Is valid CSV (unclosed quote):', isValidCsv(invalidCsv));
console.log('');

// Example 6: Get statistics
console.log('6. Get CSV statistics:');
const stats = getStats(parsed);
console.log('Statistics:', stats);
console.log('');

// Example 7: Filter rows
console.log('7. Filter rows (age > 25):');
const filtered = filterRows(parsed, (row) => parseInt(row.age as string) > 25);
console.log('Filtered rows:', filtered.rows);
console.log('');

// Example 8: Select specific columns
console.log('8. Select specific columns:');
const selected = selectColumns(parsed, ['name', 'city']);
console.log('Selected columns:', selected.headers);
console.log('Selected data:', selected.rows);
console.log('');

// Example 9: Sort by column
console.log('9. Sort by age (ascending):');
const sortedAsc = sortBy(parsed, 'age');
console.log('Sorted ascending:', sortedAsc.rows.map(r => `${r.name}: ${r.age}`));

console.log('Sort by age (descending):');
const sortedDesc = sortBy(parsed, 'age', false);
console.log('Sorted descending:', sortedDesc.rows.map(r => `${r.name}: ${r.age}`));
console.log('');

// Example 10: Add a new column
console.log('10. Add a new column:');
const withCountry = addColumn(parsed, 'country', ['USA', 'USA', 'USA', 'USA']);
console.log('New headers:', withCountry.headers);
console.log('First row with country:', withCountry.rows[0]);
console.log('');

// Example 11: Remove a column
console.log('11. Remove a column:');
const withoutCity = removeColumn(parsed, 'city');
console.log('Headers after removal:', withoutCity.headers);
console.log('');

// Example 12: Rename a column
console.log('12. Rename a column:');
const renamed = renameColumn(parsed, 'age', 'years');
console.log('Headers after rename:', renamed.headers);
console.log('First row:', renamed.rows[0]);
console.log('');

// Example 13: Merge two CSV datasets
console.log('13. Merge two CSV datasets:');
const csv1 = `
name,score
Alice,95
Bob,87
`;
const csv2 = `
name,score
Charlie,92
Diana,88
`;

const data1 = parse(csv1);
const data2 = parse(csv2);
const merged = mergeVertical(data1, data2);
console.log('Merged rows:', merged.rows);
console.log('');

// Example 14: Detect delimiter
console.log('14. Detect delimiter:');
const semicolonCsv = 'name;age;city\nJohn;30;NYC';
const tabCsv = 'name\tage\tcity\nJohn\t30\tNYC';
console.log('Detected delimiter (semicolon):', detectDelimiter(semicolonCsv));
console.log('Detected delimiter (tab):', detectDelimiter(tabCsv));
console.log('');

// Example 15: Convert to JSON
console.log('15. Convert CSV to JSON:');
const jsonData = toJson(parsed);
console.log('JSON output:', JSON.stringify(jsonData, null, 2));
console.log('');

// Example 16: Convert from JSON
console.log('16. Convert JSON to CSV:');
const jsonInput = [
  { id: 1, title: 'Task 1', completed: true },
  { id: 2, title: 'Task 2', completed: false },
  { id: 3, title: 'Task 3', completed: true },
];

const fromJsonData = fromJson(jsonInput);
console.log('Headers:', fromJsonData.headers);
console.log('CSV output:');
console.log(stringify(fromJsonData.rows));
console.log('');

// Example 17: Parse CSV without headers
console.log('17. Parse CSV without headers:');
const noHeaderCsv = 'John,30,NYC\nJane,25,LA';
const noHeaderParsed = parse(noHeaderCsv, { header: false });
console.log('Generated headers:', noHeaderParsed.headers);
console.log('Rows:', noHeaderParsed.rows);
console.log('');

// Example 18: Always quote fields
console.log('18. Always quote fields:');
const quoted = stringify(data, { alwaysQuote: true });
console.log(quoted);

console.log('=== All examples completed ===');
